stew/io2.nim

```diff
@@ -1315,3 +1368,161 @@ proc readAllChars*(pathName: string): IoResult[string] =
 proc readAllFile*(pathName: string): IoResult[seq[byte]] =
   ## Alias for ``readAllBytes()``.
   readAllBytes(pathName)
+
+proc lockFile*(handle: IoHandle, kind: LockType, offset,
+               size: int64): IoResult[void] =
+  ## Apply shared or exclusive file segment lock for file handle ``handle`` and
+  ## range specified by ``offset`` and ``size`` parameters.
+  ##
+  ## ``kind`` - type of lock (shared or exclusive). Please note that only
+  ## exclusive locks have cross-platform compatible behavior. Hovewer, exclusive
+  ## locks require ``handle`` to be opened for writing.
+  ##
+  ## ``offset`` - starting byte offset in the file where the lock should
+  ## begin. ``offset`` should be always bigger or equal to ``0``.
+  ##
+  ## ``size`` - length of the byte range to be locked. ``size`` should be always
+  ## bigger or equal to ``0``.
+  ##
+  ## If both ``offset`` and ``size`` are equal to ``0`` whole file would be
+  ## locked.
+  doAssert(offset >= 0)
+  doAssert(size >= 0)
+  when defined(posix):
+    let ltype =
+      case kind
+      of LockType.Shared:
+        cshort(posix.F_RDLCK)
+      of LockType.Exclusive:
+        cshort(posix.F_WRLCK)
+    var flockObj =
+      when sizeof(int) == 8:
+        # There is no need to perform overflow check, so we just cast.
+        FlockStruct(ltype: ltype, lwhence: cshort(posix.SEEK_SET),
+                    start: cast[int](offset), length: cast[int](size))
+      else:
+        # Currently we do not support `__USE_FILE_OFFSET64` because its
+        # Linux specific #define, and it not present in BSD systems. So
+        # on 32bit systems we do not support range locks which exceed `int32`
+        # value size.
+        if offset > int64(high(int)):
+          return err(IoErrorCode(EFBIG))
+        if size > int64(high(int)):
+          return err(IoErrorCode(EFBIG))
+        # We already made overflow check, so we just cast.
+        FlockStruct(ltype: ltype, lwhence: cshort(posix.SEEK_SET),
+                    start: cast[int](offset), length: cast[int](size))
+    while true:
+      let res = posix.fcntl(cint(handle), posix.F_SETLK, addr flockObj)
+      if res == -1:
+        let errCode = ioLastError()
+        if errCode == EINTR:
+          continue
+        else:
+          return err(errCode)
+      else:
+        return ok()
+  elif defined(windows):
+    let (lowOffsetPart, highOffsetPart, lowSizePart, highSizePart) =
+      if offset == 0'i64 and size == 0'i64:
+        # We try to keep cross-platform behavior on Windows. And we can do it
+        # because: Locking a region that goes beyond the current end-of-file
+        # position is not an error.
+        (0'u32, 0'u32, 0xFFFF_FFFF'u32, 0xFFFF_FFFF'u32)
```

Comment:
Or `uint32.high` / `not 0'u32` instead of `0xFFFF_FFFF'u32`, ensures that the number of `F` is correct. I'm fine with either.

Reply:
I'm personally do not like both `high` and `not` because its much more easier to read `0xFFFF_FFFF'u32` for anybody who are not familiar with `nim` constructs.